### PR TITLE
R in LynxKite

### PR DESCRIPTION
**Pull request by darabos**
It's working!

image

TODO:
- [x] The same for edges.
- [x] Add "derive table" and "create graph".
- [x] Docs.
- [x] Tests.
- [x] Better type support. In the screenshot `as.numeric()` is needed because Sphynx only supports int64 and float64, but `nchar()` returns int32. I don't think I want to add more types to Sphynx. Rather I think we can automatically cast to the declared type.
- [x] Make the type declarations more idiomatic. `float`, `str`, etc are from Python.
- [x] Try some fancy R package, like https://github.com/digitalcytometry/ecotyper.
- [ ] Check whether the Docker image needs any changes for this.
- [ ] Add test for Long. (Python too.)

### CODE REVIEW

**Comment by darabos** [comment]
The tests are a direct copy of ComputeInPythonTest. I'm really happy with them, they revealed a lot of mistakes and edge cases!

**Comment by darabos** [comment]
I tried a "fancy R package"! I didn't try Ecotyper. It doesn't work with graphs, and it's invoked from the command line. Instead I went for https://github.com/AlmaasLab/csdR, a coexpression scoring package. It can be used to build a graph!

image

([The code](https://gist.github.com/darabos/54d3ae5c8eb937d1b306073745053353) for reference.)

It's pretty nice mixing R and Scala and SQL in a workspace. (I didn't find an excuse to mix in Python, but it should be possible too.)

All that's left is docs! Please review!

**Comment by Gabitany** [comment]
Congrats!

**Comment by darabos** [comment]
This is ready for review! Please ignore the extra debug prints from the last commit. I'll remove them when I've fixed the issue.

It's really annoying, because it only happens on GitHub Actions. One of the Parquet files created from R seem to have two columns missing:

```
Schema mismatch for 7146c0a8-b92a-302e-b2a3-3db417d83321 (df of bc39bb3b-6d0c-3cbb-bccc-28493a6b458c (DeriveTableR(
[info] df <- df %>% mutate(age_4 = age_2 ** 2)
[info]           ,List(Field(df,name,com.lynxanalytics.biggraph.graph_api.SerializableType@556f6a27), Field(df,age,com.lynxanalytics.biggraph.graph_api.SerializableType@292c284b), Field(df,age_2,com.lynxanalytics.biggraph.graph_api.SerializableType@292c284b), Field(df,age_4,com.lynxanalytics.biggraph.graph_api.SerializableType@292c284b))))).
[info] root
[info]  |-- name: string (nullable = true)
[info]  |-- age: double (nullable = true)
[info]  |-- age_2: double (nullable = true)
[info]  |-- age_4: double (nullable = true)
[info] 
[info]  vs
[info] 
[info]  root
[info]  |-- name: string (nullable = true)
[info]  |-- age: double (nullable = true)
```

`age_2` and `age_4` are missing from the Parquet file, according to Spark. But the dataframe has these columns when I write it out from R. Hard to tell who is right. I can't take a look at the file. The only clue I have is that On GitHub actions it's running with `r-arrow` version 9.0.0, while locally I have 8.0.1. Funny thing is, I can't install 9.0.0 locally because other packages (cugraph) depend on older versions.

Tomorrow is a working day in Hungary. (So we don't have to work on the 31st.) I may give #295 a try. Otherwise it's a time sink to troubleshoot these things.

**Comment by darabos** [comment]
This has an issue where r-arrow 9.0.0 preserves the metadata from the Spark-written Parquet file. This causes Spark not to see the newly added columns. I have a fix for this in #296 but I made it to the fix only after replacing the whole CI system.

So I'll just merge this and then #296. The two together make everything work well, I think.

**Comment by lacca0** [comment]
image
We need to add an info pop-up window.

**Comment by lacca0** [comment]
Снимок экрана от 2022-11-10 15-23-45
We need to add an info pop-up in the dropdown menu

**Comment by lacca0** [comment]
Also please add KITE_ALLOW_R explanation in kiterc_template. Thanks!

**Comment by darabos** [comment]
> We need to add an info pop-up in the dropdown menu.

Fixed.

> Also please add KITE_ALLOW_R disambiguation in kiterc_template. Thanks!

Done.

> image We need to add an info pop-up window.

I'm not sure what you mean. If you click on the output you also get the error in a popup:

image

But that's basically it. If the administrator has disabled it, there's nothing the user can do about it.

Or do you mean it should say "assertion failed"? I'll look into that.

**Comment by lacca0** [comment]
You are replying to the wrong screenshot, I accidentally added that one and changed it later (sorry). I hope the new one will be more explanatory.

**Comment by lacca0** [comment]
>I'm not sure what you mean.
I mean adding something like this in conf/kiterc_template:
```
# Allow executing R code
export KITE_ALLOW_R=yes
```

**Comment by darabos** [comment]
> I'm not sure what you mean. If you click on the output you also get the error in a popup:
> 
> image
> 
> But that's basically it. If the administrator has disabled it, there's nothing the user can do about it.
> 
> Or do you mean it should say "assertion failed"? I'll look into that.

Okay, now I know it was just the wrong screenshot. But I've fixed it anyway. Look how much nicer it is!

image

**Comment by darabos** [comment]
Thanks a lot for the comments! Thanks for the thorough testing too! I don't know if you've caught _all_ my mistakes, but you caught a good number of them for sure.

**Review by lacca0 — APPROVED** [review]

File: sphynx/r/derive_table.r [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+code <- params[["code"]]
+eval(parse(text = code))
+print("USER CODE FINISHED")
+
+# Save outputs.
+print(params[["outputFields"]])
+for (t in names(outputs)) {
+    print(t)
+    fields <- params[["outputFields"]]
+    fs <- fields[fields$parent == t, ]
```
[the remark this answers is not in the document]
Comment (darabos) [reply sphynx/r/derive_table.r:20]: > Please comment on why the comma is needed here.

Looks like this is just basic R. See https://stackoverflow.com/questions/1686569/filter-data-frame-rows-by-a-logical-condition.

I like the `filter(df, ...)` syntax more so I tried switching to that. `fs <- filter(fields, parent == parent)` But here the condition is always true! Filter has no way of knowing I was thinking of the column on the left side and the local variable on the right side!

That brings me to https://stackoverflow.com/questions/40169949/filter-dataframe-using-global-variable-with-the-same-name-as-column-name which says _"just use `df[df$y==y,]` instead of `dplyr`."_

I guess when we get more familiar with R it won't be an issue.

**Comment by darabos** [comment]
Thanks! I've fixed the comments you commented on and sorted out the TODOs I forgot about earlier. Debug prints are gone.
